Mathlib/GameTheory/Auction/Secondprice.lean

```diff
@@ -0,0 +1,208 @@
+attribute [instance] Auction.hF Auction.hI
+
+namespace Auction
+
+variable {a : Auction} (b : a.I → ℝ)
+
+/-!### Helper Functions and Definitions -/
+/-- Computes the highest bid given a bidding function `b`. -/
+@[simp]
+def maxBid : ℝ := Finset.sup' Finset.univ Finset.univ_nonempty b
+
+/-- There exists a participant `i` whose bid equals the highest bid. -/
+lemma exists_maxBid : ∃ i : a.I, b i = a.maxBid b := by
+   obtain ⟨i, _, h2⟩ := Finset.exists_mem_eq_sup' Finset.univ_nonempty b
+   exact ⟨i, symm h2⟩
+
+/-- winner: The participant with the highest bid. -/
+noncomputable def winner : a.I := Classical.choose (exists_maxBid b)
+
+/-- The bid of the winner equals the highest bid. -/
+lemma bid_winner_eq_maxBid : b (winner b) = maxBid b := Classical.choose_spec (exists_maxBid b)
+
+/-- The bid of the winner is always greater than or equal to the bids of all others.-/
+lemma bid_le_bid_winner (j : a.I) : b j ≤ b (winner b) := by
+  rw [bid_winner_eq_maxBid b]
+  exact Finset.le_sup' b (Finset.mem_univ j)
+
+/-- If `i`'s bid is higher than all other bids, then `i` is the winner. -/
+lemma eq_winner_of_bid_gt (i : a.I) (H : ∀ j , j ≠ i → b j < b i) : i = winner b := by
+  contrapose! H
+  exact ⟨winner b, H.symm, bid_le_bid_winner b i⟩
+
+/-- A strategy is dominant if bidding `bi` ensures that
+`i`'s utility is maximized relative to any other bids `b'` where `b i = bi`. -/
+def dominant (utility : (a.I → ℝ) → a.I → ℝ) (i : a.I) (bi : ℝ) : Prop :=
+  ∀ b, utility b i ≤ utility (Function.update b i bi) i
+
+/-! ### Proofs and Lemmas -/
+
+namespace Firstprice
+
+/-- Computes the utility for a first price auction where the winner pays their bid. -/
+noncomputable def utility (i : a.I) : ℝ := if i = winner b then a.v i - b i else 0
+
+/-- If `i` is the winner in a first price auction, utility is their valuation minus their bid. -/
+lemma utility_winner (i : a.I) (H : i = winner b) : utility b i = a.v i - b i := by
+  rw [H]
+  simp only [utility, if_true]
+
+/-- If `i` is not the winner in a first price auction, their utility is 0. -/
+lemma utility_loser (i : a.I) (H : i ≠ winner b) : utility b i = 0 := by
+  rw [utility]
+  simp only [H, if_false]
+
+/-- Shows that there is no dominant strategy in a first price auction for any `i` and bid `bi`. -/
+theorem firstprice_auction_has_no_dominant_strategy (i : a.I) (bi : ℝ) :
+    ¬ dominant utility i bi := by
+  rw [dominant, not_forall]
+  use Function.update (fun _ ↦ (bi - 2)) i (bi - 1)
+  rw [utility, utility, if_pos (eq_winner_of_bid_gt _ i _), if_pos (eq_winner_of_bid_gt _ i _)]
+  <;> intros <;> simp [*]
+
+end Firstprice
+
+/-- `maxBidExcluding i` is the maximal bid of all participants but `i`. -/
+noncomputable def maxBidExcluding (i : a.I) : ℝ := Finset.sup' (Finset.erase Finset.univ i)
+(Finset.Nontrivial.erase_nonempty (Finset.univ_nontrivial)) b
+
+/--The second highest bid: the highest bid excluding the winner’s bid.-/
+noncomputable def Secondprice : ℝ := maxBidExcluding b (winner b)
```

Comment:
I think this should be lower camel case, unless you're viewing it as one word (https://leanprover-community.github.io/contribute/naming.html).
```suggestion
noncomputable def secondPrice : ℝ := maxBidExcluding b (winner b)
```